core/src/main/scala/kafka/server/metadata/QuotaCache.scala

```diff
@@ -0,0 +1,242 @@
+sealed trait CacheIndexKey
+case object DefaultUser extends CacheIndexKey
+case class SpecificUser(user: String) extends CacheIndexKey
+case object DefaultClientId extends CacheIndexKey
+case class SpecificClientId(clientId: String) extends CacheIndexKey
+case object DefaultIp extends CacheIndexKey
+case class SpecificIp(ip: String) extends CacheIndexKey
+
+
+// Different types of matching constraints
+sealed trait QuotaMatch
+case class ExactMatch(entityName: String) extends QuotaMatch
+case object DefaultMatch extends QuotaMatch
+case object TypeMatch extends QuotaMatch
+
+
+class QuotaCache {
+  type QuotaCacheIndex = mutable.HashMap[CacheIndexKey, mutable.HashSet[QuotaEntity]]
+
+  // A cache of the quota entities and their current quota values
+  private val quotaCache = new mutable.HashMap[QuotaEntity, mutable.Map[String, Double]]
+
+  // An index of user or client to a set of corresponding cache entities. This is used for flexible lookups
+  private val userEntityIndex = new QuotaCacheIndex
+  private val clientIdEntityIndex = new QuotaCacheIndex
+  private val ipEntityIndex = new QuotaCacheIndex
+
+  private val lock = new ReentrantReadWriteLock()
+
+  /**
+   * Return quota entries for a given filter. These entries are returned from an in-memory cache and may not reflect
+   * the latest state of the quotas according to the controller.
+   *
+   * @param quotaFilter       A quota entity filter
+   * @return                  A mapping of quota entities along with their quota values
+   */
+  def describeClientQuotas(quotaFilter: ClientQuotaFilter): Map[QuotaEntity, Map[String, Double]] = inReadLock(lock) {
+
+    // Do some preliminary validation of the filter types and convert them to correct QuotaMatch type
+    val entityFilters: mutable.Map[String, QuotaMatch] = mutable.HashMap.empty
+    quotaFilter.components().forEach(component => {
+      val entityType = component.entityType()
+      if (entityFilters.contains(entityType)) {
+        throw new InvalidRequestException(s"Duplicate ${entityType} filter component entity type")
+      } else if (entityType.isEmpty) {
+        throw new InvalidRequestException("Unexpected empty filter component entity type")
+      } else if (!ClientQuotaEntity.isValidEntityType(entityType)) {
+        throw new UnsupportedVersionException(s"Custom entity type ${entityType} not supported")
+      }
+
+      // A present "match()" is an exact match on name, an absent "match()" is a match on the default entity,
+      // and a null "match()" is a match on the entity type
+      val entityMatch = if (component.`match`() != null && component.`match`().isPresent) {
+        ExactMatch(component.`match`().get())
+      } else if (component.`match`() != null) {
+        DefaultMatch
+      } else {
+        TypeMatch
+      }
+      entityFilters.put(entityType, entityMatch)
+    })
+
+    if (entityFilters.isEmpty) {
+      // TODO return empty or throw exception here?
+      return Map.empty
```

Comment:
Does this code execute if a client tries to describe everything -- and is that legal to do?  I'm not sure.

Reply:
The `kafka-configs.sh` script requires at least one entity type filter, so we can't reach this code via the CLI. However, a client could send an RPC message with no filters. Looks like the current ZK logic prohibits a describe without at least one entity type filter, so we'll do the same.